Inject `mirrord-key` into all forwarded messages (#268)
metalbear-co/docs: docs/sharing-the-cluster/queue-splitting.md

Reviewer: ```suggestion

```

Reviewer: ```suggestion
## Choose your queue service
```

Reviewer: ```suggestion
## How It Works
```

| RabbitMQ | Metadata | Message header |
| Apache Kafka | Metadata | Message header |
| Azure Service Bus | Metadata | Application property |
| Temporal | Metadata | Activity task header |
| BullMQ | JSON payload | Job `data` object |
| Redis Pub/Sub | JSON payload | Message payload |

For BullMQ and Redis Pub/Sub, where the key lives in the JSON payload itself: a payload that is not a JSON object is forwarded unchanged, and an existing `mirrord-key` in the message is never overwritten.

For Temporal, only the activity task is stamped. Workflow tasks are not, because their header lives in workflow history that the worker replays and validates against the server.

{% hint style="info" %}
Injection is enabled through the `operator.injectSessionKeyHeader` setting in the [mirrord-operator Helm chart](https://github.com/metalbear-co/charts/blob/main/mirrord-operator/values.yaml).
{% endhint %}

{% hint style="info" %}
The key is not injected in copy-target mode. A copy-target session runs against its own dedicated copy of the workload, so no shared consumer needs to tell sessions apart, and messages routed to the copy are delivered unchanged.
{% endhint %}

> ### Setting a Filter for a mirrord Run

Reviewer: ```suggestion
## Setting a Filter for a mirrord Run
```

2. In case of SQS splitting, deployed targets will keep reading from the temporary queues as long as their temporary queues have unconsumed messages.
3. For Google Cloud Pub/Sub, the operator creates temporary topics and subscriptions. The target workload's subscription environment variable is patched to read from a temporary subscription, while the operator drains the original subscription and forwards messages through temporary topics.

> ### Session Key Header

Reviewer: ```suggestion
## Session Key Header
```

When your operator has session key header injection enabled, every message the operator routes to your session is stamped with a `mirrord-key` carrying your session key, so your local application can tell which mirrord session a message belongs to. Only the copy delivered to your session is stamped; the message the deployed application receives is never modified.

Where the key is placed depends on the queue service. Services with a metadata channel carry the key there; the rest carry it inside the JSON payload.

| Queue service | Carrier | Location of `mirrord-key` |
| --- | --- | --- |
| Amazon SQS | Metadata | Message attribute |
| Google Cloud Pub/Sub | Metadata | Message attribute |
| RabbitMQ | Metadata | Message header |
| Apache Kafka | Metadata | Message header |
| Azure Service Bus | Metadata | Application property |
| Temporal | Metadata | Activity task header |
| BullMQ | JSON payload | Job `data` object |
| Redis Pub/Sub | JSON payload | Message payload |

> For a JSON-payload service, a payload that is not a JSON object is forwarded unchanged, and an existing `mirrord-key` in the message is never overwritten. For Temporal, only the activity task is stamped; workflow tasks are not, because their header lives in workflow history that the worker replays and validates against the server.

Reviewer: ```suggestion
For BullMQ and Redis Pub/Sub, where the key lives in the JSON payload itself: a payload that is not a JSON object is forwarded unchanged, and an existing `mirrord-key` in the message is never overwritten.

For Temporal, only the activity task is stamped. Workflow tasks are not, because their header lives in workflow history that the worker replays and validates against the server.
```

{% hint style="info" %}
Injection is enabled through the `operator.injectSessionKeyHeader` setting in the [mirrord-operator Helm chart](https://github.com/metalbear-co/charts/blob/main/mirrord-operator/values.yaml).
{% endhint %}

> {% hint style="warning" %}

Reviewer: shouldnt be a warning imo
